app/modules/asset/service.server.ts

```diff
@@ -101,25 +102,59 @@ type AssetWithInclude<T extends Prisma.AssetInclude | undefined> =
 export async function getAsset<T extends Prisma.AssetInclude | undefined>({
   id,
   organizationId,
+  userOrganizations,
   include,
 }: Pick<Asset, "id"> & {
   organizationId: Asset["organizationId"];
+  userOrganizations: Pick<UserOrganization, "organizationId">[];
   include?: T;
 }): Promise<AssetWithInclude<T>> {
   try {
+    const otherOrganizationIds = userOrganizations.map(
+      (org) => org.organizationId
+    );
+
     const asset = await db.asset.findFirstOrThrow({
-      where: { id, organizationId },
+      where: {
+        OR: [
+          { id, organizationId },
+          { id, organizationId: { in: otherOrganizationIds } },
+        ],
+      },
       include: { ...include },
     });
 
+    /* User is accessing the asset in the wrong organization. In that case we need special 404 handling. */
+    if (
+      asset.organizationId !== organizationId &&
+      otherOrganizationIds.includes(asset.organizationId)
+    ) {
+      throw new ShelfError({
+        cause: null,
+        title: "Asset not found",
+        message: "",
+        additionalData: {
+          type: "asset-from-other-org",
+          assetOrganization: userOrganizations.find(
+            (org) => org.organizationId === asset.organizationId
+          ),
+        },
+        label,
+      });
+    }
+
     return asset as AssetWithInclude<T>;
   } catch (cause) {
     throw new ShelfError({
       cause,
       title: "Asset not found",
       message:
         "The asset you are trying to access does not exist or you do not have permission to access it.",
-      additionalData: { id, organizationId },
+      additionalData: {
+        id,
+        organizationId,
+        ...(cause instanceof ShelfError ? cause.additionalData : {}),
```

Comment:
We have a helper function called `isLikeShelfError`. You can search the code to see how we use it. Please use that instead of checking it yourself.